File: Extras/mocha/test/unium.js
// a class to make the websocket interface a little more test friendly

const chai      = require('chai');
const WebSocket = require( 'ws' );

const { expect } = chai;

module.exports = class UniumHelper {

  constructor() {
    this.ws      = null;
    this.next_id = 1;
    this._on     = {};
    this._once   = {};
  }

  connect( uri ) {
    return new Promise( (resolve,reject) => {
      this.disconnect();
      this.ws = new WebSocket( uri );
      this.ws.on( 'open', resolve );
      this.ws.on( 'error', reject );
      this.ws.on( 'message', (m) => this.onMessage(m) );
    });
  }

  disconnect() {
    if( this.ws ) {
      this.ws.close();
      this.ws = null;
    }
  }


  onMessage( msg ) {

    const data = JSON.parse( msg );
    const {id} = data;

    if(!id) {
      return;
    }

    if( id in this._on ) {
      this._on[ id ].forEach( fn => fn( data ) );
    }

    if( id in this._once ) {
      this._once[ id ].forEach( fn => fn( data ) );
      delete this._once[ id ];
    }
  }
  
  wait_for( msg, timeout = 2 ) {
    return new Promise( (resolve,reject) => {
      const timeoutHandle = setTimeout( () => reject( 'timeout out' ), timeout * 1000 );
      this.once( msg, (data) => {
        clearTimeout( timeoutHandle );
        resolve( data );
      });
    });
  }

  async send( uri, name ) {

    const msg = {
      id: name || `m${this.next_id++}`,
      q: uri
    };

    this.ws.send( JSON.stringify( msg ) )

    return msg.id;
  }

  async get( uri, timeout ) {

    const mid = await this.send( uri );
    const msg = await this.wait_for( mid, timeout );

    expect( msg ).to.not.be.null;

    msg.should.have.property( 'id' );
    msg.should.not.have.property( 'error' );
    msg.should.have.property( 'data' );
    msg.id.should.equal( mid );

    return msg.data;
  }

  async bind( uri, name ) {

    const mid = name || uri.split('.').pop();

    await this.send( `/bind${uri}`, mid );
    const res = await this.wait_for( mid );

    expect( res ).to.not.be.null;
    res.should.have.property( 'id' );
    res.should.not.have.property( 'error' );
    res.should.have.property( 'info' );

    return res;
  }

  on( event, fn ) {
    if( event in this._on ) {
      this._on[ event ].push( fn );
    } else {
      this._on[ event ] = [ fn ];
    }
  }

  once( event, fn ) {
    if( event in this._once ) {
      this._once[ event ].push( fn );
    } else {
      this._once[ event ] = [ fn ];
    }
  }
}

File: Extras/nunit/Unium.cs
using Newtonsoft.Json.Linq;
using NUnit.Framework;
using System;
using System.Collections.Generic;
using System.Net.WebSockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Unium.Helpers
{
    public class WebsocketHelper : IDisposable
    {
        private int             mNextID = 1;
        private ClientWebSocket mWS     = null;

        private Dictionary<string, List<Action<JToken>>> mOn;
        private Dictionary<string, List<Action<JToken>>> mOnce;


        public async Task Connect( string uri )
        {
            mWS = new ClientWebSocket();

            using( var ct = new CancellationTokenSource( 2000 ) )
            {
                var taskConnect = mWS.ConnectAsync( new Uri( uri ), ct.Token );

                Assert.True(
                    ( mWS.State == WebSocketState.None ) ||
                    ( mWS.State == WebSocketState.Connecting ) ||
                    ( mWS.State == WebSocketState.Open )
                );

                await taskConnect;
            }

            Assert.AreEqual( WebSocketState.Open, mWS.State );

            mNextID = 1;
            mOn     = new Dictionary<string, List<Action<JToken>>>();
            mOnce   = new Dictionary<string, List<Action<JToken>>>();
        }


        public async Task Disconnect()
        {
            await mWS.CloseAsync( WebSocketCloseStatus.NormalClosure, string.Empty, CancellationToken.None );
            mOn   = null;
            mOnce = null;
        }


        public void Dispose()
        {
            if( mWS != null )
            {
                mWS.Dispose();
            }
        }


        private void OnMessage( string id, JToken msg )
        {
            if( string.IsNullOrWhiteSpace( id ) )
            {
                return;
            }

            if( mOn.ContainsKey( id ) )
            {
                mOn[ id ].ForEach( h => h( msg ) );
            }

            if( mOnce.ContainsKey( id ) )
            {
                mOnce[ id ].ForEach( h => h( msg ) );
                mOnce.Remove( id );
            }
        }


        public void On( string id, Action<JToken> callback )
        {
            if( mOn.ContainsKey( id ) == false )
            {
                mOn.Add( id, new List<Action<JToken>>() );
            }

            mOn[ id ].Add( callback );
        }


        public void Once( string id, Action<JToken> callback )
        {
            if( mOnce.ContainsKey( id ) == false )
            {
                mOnce.Add( id, new List<Action<JToken>>() );
            }

            mOnce[ id ].Add( callback );
        }


        public string Send( string q, string name = null )
        {
            var id    = name ?? $"m{mNextID++}";
            var query = Encoding.ASCII.GetBytes( $@"{{""id"":""{id}"",""q"":""{q}""}}" );

            mWS.SendAsync( new ArraySegment<byte>( query ), WebSocketMessageType.Text, true, CancellationToken.None );

            return id;
        }


        private async Task<JToken> WaitForTask( string id )
        {
            while( true )
            {
                var buffer  = new byte[ 2048 ];
                var segment = new ArraySegment<byte>( buffer, 0, buffer.Length );

                var recv = await mWS.ReceiveAsync( segment, CancellationToken.None );
                Assert.AreEqual( WebSocketMessageType.Text, recv.MessageType );

                var data = Encoding.UTF8.GetString( buffer );
                var msg  = JToken.Parse( data );

                var mid = msg.Value<string>( "id" );

                OnMessage( mid, data );

                if( id == null || mid == id )
                {
                    return msg;
                }
            }
        }

        public async Task<JToken> WaitFor( string id, float timeout = 2.0f )
        {
            var task = WaitForTask( id );

            if( await Task.WhenAny( task, Task.Delay( (int) ( timeout * 1000.0f ) ) ) != task )
            {
                Assert.Fail( "Timeout" );
            }

            return task.Result;
        }

        public async Task<JToken> Get( string uri, float timeout = 2.0f )
        {
            var id  = Send( uri );
            var res = await WaitFor( id, timeout );
            return res.Value<JToken>( "data" );
        }


        public async Task Bind( string uri, string name = null )
        {
            var parts = uri.Split( new char[] { '.' } );
            var eventName = parts[ parts.Length - 1 ];

            string id = name ?? eventName;

            Send( $"/bind{uri}", id );

            dynamic res = await WaitFor( id );

            Assert.IsNull( (string) res.error );
            Assert.AreEqual( id, (string) res.id );
            Assert.AreEqual( "bound", (string) res.info );
        }
    }
}
